### PR TITLE
kubeadm: Add support for using an external CA whose key is never stored in the cluster

### LINKED ISSUE #280: Make it easy to run an other certificates controller
## Is this a BUG REPORT or FEATURE REQUEST?

Choose one: BUG REPORT or FEATURE REQUEST

FEATURE REQUEST

It would be nice if we made it easy to deploy your own certificates controller instead of relying on the one that's built-in into the core controller-manager.

If the user opts in to running a separate certificates signing controller, the availability of `ca.key` on disk will be optional _if and only if_ the user specifies the rest of the required files, namely:
 - ca.crt
 - apiserver.{crt,key}
 - apiserver-kubelet-client.{crt,key}
 - admin.conf
 - kubelet.conf
 - controller-manager.conf
 - scheduler.conf

This assumes the CSR approval controller will still be run, so it's only the CSR signing controller that's deactivated. This can be done by just omitting the `cluster-signing-cert-file` and `cluster-signing-key-file` flags to the controller-manager.

Combined with kubelet certification rotation, all kubelets (including the master) will have a client/server cert/key pair signed by the external CA without ever having the CA key on disk.

kubeadm changes: The certificates controller can be added just by a `kubectl apply -f` post-installation, the only thing that needs to be done is to inactivate the default CSR signing controller.
We should document how to do this in detail.

Is there anything I missed that has to be done for this to work?

Thoughts?
@kubernetes/sig-auth-feature-requests @mikedanese @jcbsmpsn

### PR DESCRIPTION
**Pull request by nckturner**
We allow a kubeadm user to use an external CA by checking to see if ca.key is missing and skipping cert checks and kubeconfig generation if ca.key is missing.  We also pass an empty arg --cluster-signing-key-file="" to kube controller manager so that the csr signer doesn't start. 

**What this PR does / why we need it**:

This PR allows the kubeadm certs phase and kubeconfig phase to be skipped if the ca.key is missing but all other certs are present.  

**Which issue this PR fixes** : 

Fixes kubernetes/kubeadm/issues/280

**Special notes for your reviewer**:

@luxas @mikedanese @fabriziopandini 

**Release note**:

```release-note
kubeadm: Add support for using an external CA whose key is never stored in the cluster
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/init.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: nit: `fmt.Println`
nit: `No ca.key detected, but all other files generated from it are available, so using external CA mode. Won't generate certs nor kubeconfig files` (or something like that)

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (luxas) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: for clarity; I'd always set `cluster-signing-key-file` and `cluster-signing-cert-file` in `defaultArguments`
and in this special case; unset both (although one only is technically needed, it's safest to reset both of them

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (luxas) [inline cmd/kubeadm/app/phases/certs/certs.go]: What about `sa.pub`? can we make a case for that one as well?

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (luxas) [inline cmd/kubeadm/app/phases/certs/certs.go]: Please add a comment for all these

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (luxas) [inline cmd/kubeadm/app/phases/certs/certs.go]: And unit test all these new funcs you are creating.

**Review by luxas — COMMENTED** [review]
Basically LGTM; some asks and nits + unit (which I know you're working on already :+1:)

Thanks for this contribution!

**Comment by luxas** [comment]
From the verify job:
```
I0817 11:02:41.808] Run ./hack/update-bazel.sh
...
W0817 11:20:11.506] Errors from golint:
W0817 11:20:11.507] cmd/kubeadm/app/phases/certs/certs.go:162:1: exported function ValidateCACert should have comment or be unexported
W0817 11:20:11.510] cmd/kubeadm/app/phases/certs/certs.go:178:1: exported function ValidateCACertAndKey should have comment or be unexported
W0817 11:20:11.510] cmd/kubeadm/app/phases/certs/certs.go:191:1: exported function ValidateSignedCert should have comment or be unexported
W0817 11:20:11.511] cmd/kubeadm/app/phases/certs/certs.go:214:1: exported function ValidatePrivateKey should have comment or be unexported
```
Please fix.

**Comment by luxas** [comment]
This definitely deserves a release note :smile:

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (mikedanese) [inline cmd/kubeadm/app/phases/certs/certs.go]: don't start go variables with capitals. caBaseName...

**Comment by fabriziopandini** [comment]
:+1

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (mikedanese) [inline cmd/kubeadm/app/phases/certs/certs.go]: make this table driven. https://play.golang.org/p/UJY4McemtD

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (mikedanese) [inline cmd/kubeadm/app/phases/certs/certs.go]: raw prints without linebreaks are going to look pretty funky in the output.

File: cmd/kubeadm/app/cmd/init.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/init.go]: According to the issue, it is expected to have also required kubeconfig files on disk. Is kubeadm going to test this precondition?

Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: Yes, that should also be checked, good catch

File: cmd/kubeadm/app/cmd/init.go
Comment (nckturner) [inline cmd/kubeadm/app/cmd/init.go]: Fixed

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (nckturner) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Makes sense, fixed.

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (nckturner) [inline cmd/kubeadm/app/phases/certs/certs.go]: Can you explain what you are suggesting a little more?  Do you mean this entire function?

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (nckturner) [inline cmd/kubeadm/app/phases/certs/certs.go]: I added a case, to superficially check for its existence, tell me what you think.

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (nckturner) [inline cmd/kubeadm/app/phases/certs/certs.go]: Done

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (nckturner) [inline cmd/kubeadm/app/phases/certs/certs.go]: Fixed

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (nckturner) [inline cmd/kubeadm/app/phases/certs/certs.go]: Done, though tell me if I should refactor them into table driven tests.

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (nckturner) [inline cmd/kubeadm/app/phases/certs/certs.go]: Fixed

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (luxas) [inline cmd/kubeadm/app/phases/certs/certs.go]: Table driven tests are common practice, please write such ones.
Nearly all unit tests in `cmd/kubeadm` are table driven

File: cmd/kubeadm/app/cmd/init.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: I don't think we should print anything here, this is the "normal" state anyway...

File: cmd/kubeadm/app/cmd/init.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: err seems to be ignored here?

File: cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go
Comment (luxas) [inline cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go]: Any reason to not use := here?

File: cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go
Comment (luxas) [inline cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go]: nit on style, this can be moved out of the if clause and you can check for `!ok`, return in that case, otherwise just use `:=` afterwards.

Generally I'm trying to minimize the amount of indentation in funcs and avoid if/else when possible. This is still a nit though ;)

**Review by luxas — COMMENTED** [review]
Left some nits, please fix at least most of them. Other than that, LGTM

File: cmd/kubeadm/app/cmd/init.go
Comment (nckturner) [reply cmd/kubeadm/app/cmd/init.go]: I'd say its not an error at this step, so I changed it to be a string representing the reason that we are not in "Using External CA".

File: cmd/kubeadm/app/cmd/init.go
Comment (nckturner) [reply cmd/kubeadm/app/cmd/init.go]: Removed the print, if someone wants to see why external CA mode is not active they would have to add print statements to this step.  (a logger with verbosity levels would be nice but it seems to be not used in kubeadm)

File: cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go
Comment (nckturner) [reply cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go]: Fixed

File: cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go
Comment (nckturner) [reply cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go]: Fixed.

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (luxas) [inline cmd/kubeadm/app/phases/certs/certs.go]: I actually preferred returning an error like you did in the earlier version, but I'm not blocking this PR on that

File: cmd/kubeadm/app/phases/certs/certs.go
Comment (luxas) [inline cmd/kubeadm/app/phases/certs/certs.go]: Not needed for this PR; but in some future PR we should make sure the public key in the cert and key match each other (we have other places that do similar functionality like this so...)

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (luxas) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Gah, should have spotted this earlier, I realize `delete(defaultArguments, "cluster...")` is probably preferable over this method... WDYT?

It would be great to get an unit test for this as well if possible...

**Review by luxas — COMMENTED** [review]
WDYT about the `delete(...` comment?

File: cmd/kubeadm/app/phases/controlplane/manifests.go
Comment (nckturner) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: The nice thing about setting to "" is that (in controller manager logs) you get:

> Skipping "csrsigning"

instead of: 

> Failed to start certificate controller: error reading CA cert file "/etc/kubernetes/ca/ca.pem": open /etc/kubernetes/ca/ca.pem: no such file or directory

when using `delete()` which I think is slightly better, but I will switch them if you prefer.

Comment (luxas) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: With that context (thanks for reminding me about the UX), it makes sense to pass `""`

**Review by luxas — APPROVED** [review]
Thanks for this!

/lgtm

**Comment by nckturner** [comment]
@luxas @mikedanese I squashed my commits, can I get another lgtm?

**Comment by fabriziopandini** [comment]
🎉 well done!

**Pull request merged**
